### ISSUE TITLE
Make model architecture compatible with deployment

### ISSUE THREAD

**Issue by ziw-liu** [mehta-lab/microDL#214]
Currently the torch modules inherit the design paradigm where dynamic business logic, such as input check, is incorporated liberally in the models themselves. This kind of behavior can cause issues in deployment (e.g. https://github.com/mehta-lab/microDL/issues/208).

Since there is interest in deploying the torch models on edge devices for real-time inference, care should be taken to either reduce these operations or use specific techniques to implement them (e.g. TorchScript). Here is a relevant [torch docs page](https://pytorch.org/docs/stable/onnx.html#tracing-vs-scripting) about model implementation when converting to ONNX is desired.

**Comment by Christianfoley** [comment]
Creating a list of opset incompatibilities (see [official list](https://pytorch.org/docs/stable/onnx_supported_aten_ops.html)) between our model and ONNX runtime. Will update periodically.
- `aten::_convolution_mode` **FIX:** remove `padding='same' # or 'valid'` parameter from all convolution layers and implement padding logic manually.

**Comment by Christianfoley** [comment]
Will be difficult to test the model on slurm reservations [(see this issue)](https://github.com/microsoft/onnxruntime/issues/8313#issuecomment-1051481689).

**Comment by ziw-liu** [comment]
> Will be difficult to test the model on slurm reservations [(see this issue)](https://github.com/microsoft/onnxruntime/issues/8313#issuecomment-1051481689).

Does this persist if you get an exclusive node?

**Comment by ziw-liu** [comment]
> * `aten::feature_dropout` conversion to onnx requires setting an inference shape, which disagrees with the call behavior in pytorch. Not sure how to fix this.

Not sure I'm following here. Usually we don't want drop out (and maybe also batchnorm etc) during inference, and a properly written model should turn off these layers when `model.eval()` or `torch.inference_mode()`.

Assuming the above assumption is true, what we actually need to do is to replace stuff like:
https://github.com/mehta-lab/microDL/blob/1e936662dc708487692d2036d6c050e307cea20f/micro_dl/torch_unet/networks/layers/ConvBlock3D.py#L98-L102
with a torch [ModuleList](https://github.com/mehta-lab/microDL/blob/1e936662dc708487692d2036d6c050e307cea20f/micro_dl/torch_unet/networks/layers/ConvBlock3D.py#L98-L102).

I can see that other plain lists of modules are manually registered with:
https://github.com/mehta-lab/microDL/blob/1e936662dc708487692d2036d6c050e307cea20f/micro_dl/torch_unet/networks/layers/ConvBlock3D.py#L328-L339

This can be error-prone. @Christianfoley can we just always use `ModuleList` instead?

**Comment by ziw-liu** [comment]
@Christianfoley Also I'm puzzled by [the initialization of the dropout layers](https://github.com/mehta-lab/microDL/blob/1e936662dc708487692d2036d6c050e307cea20f/micro_dl/torch_unet/networks/layers/ConvBlock3D.py#L98-L102):
```py
nn.Dropout3d(int(self.dropout))
```
This will always be `nn.Dropout3d(p=0)` given:
https://github.com/mehta-lab/microDL/blob/1e936662dc708487692d2036d6c050e307cea20f/micro_dl/torch_unet/networks/Unet25D.py#L57-L59

So it's never effectively used anyway?

**Comment by ziw-liu** [comment]
> * `aten::feature_dropout` conversion to onnx requires setting an inference shape, which disagrees with the call behavior in pytorch. Not sure how to fix this.

Fixed by d1b6ec75edd19e1bcdafa6fceb35449472a0bbdc

**Comment by Christianfoley** [comment]
> This can be error-prone. @Christianfoley can we just always use ModuleList instead?

Is this error prone? I see no issue with the use of a ModuleList but I believe their behavior is the same (PyTorch calls `add_module` on the backend in the [ModuleList source](https://pytorch.org/docs/master/_modules/torch/nn/modules/container.html#ModuleList) I think)

If it's actively causing issues it's worth a change, though.

**Comment by ziw-liu** [comment]
> Is this error prone?

This means that people need to manually register modules, and as exemplified here, we don't always remember to do so and introduce bugs. Also 'it calls it for you' is just better IMO because why do it ourselves if we don't have to?

**Comment by ziw-liu** [comment]
Closed by #234
